File: service/app/src/main/java/io/miragon/blueprint/domain/leasing/ApplicationInvalidException.java
package io.miragon.blueprint.domain.leasing;

/**
 * Raised while validating an application that cannot proceed (e.g. no income to cover the rate).
 * The inbound Operaton adapter translates this into the BPMN error {@code applicationInvalid},
 * which the process catches on the validate task's boundary event.
 */
public class ApplicationInvalidException extends RuntimeException {

    private final ApplicationId applicationId;
    private final String reason;

    public ApplicationInvalidException(ApplicationId applicationId, String reason) {
        super("Application " + applicationId.value() + " is invalid: " + reason);
        this.applicationId = applicationId;
        this.reason = reason;
    }

    public ApplicationId getApplicationId() {
        return applicationId;
    }

    public String getReason() {
        return reason;
    }
}

File: service/app/src/main/kotlin/io/miragon/blueprint/domain/leasing/ApplicationInvalidException.kt
package io.miragon.blueprint.domain.leasing

/**
 * Raised while validating an application that cannot proceed (e.g. no income to cover the rate).
 * The inbound Operaton adapter translates this into the BPMN error `applicationInvalid`,
 * which the process catches on the validate task's boundary event.
 */
class ApplicationInvalidException(
    val applicationId: ApplicationId,
    val reason: String,
) : RuntimeException("Application ${applicationId.value} is invalid: $reason")
